libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,880 @@
+    * The only restrictions on that field are that it is serializable to a binary representation.
+    * Convenience functions exist to handle most of the primitive types as well as some more complex types, and are
+    * used automatically where possible.
+    */
+   class kv_index {
+      class iterator {
+      public:
+         iterator(eosio::name contract_name, uint32_t itr, kv_it_stat itr_stat, kv_index* idx) :
+                  contract_name{contract_name}, itr{itr}, itr_stat{itr_stat}, idx{idx} {}
+
+         iterator(iterator&& other) :
+            contract_name(std::move(other.contract_name)),
+            itr(std::exchange(other.itr, 0)),
+            itr_stat(std::move(other.itr_stat)),
+            idx(std::exchange(other.idx, nullptr))
+         {}
+
+         ~iterator() {
+            if (itr) {
+               internal_use_do_not_use::kv_it_destroy(itr);
+            }
+         }
+
+         iterator& operator=(iterator&& other) {
+            contract_name = std::move(other.contract_name);
+            itr = std::exchange(other.itr, 0);
+            itr_stat = std::move(other.itr_stat);
+            idx = std::exchange(other.idx, nullptr);
+
+            return *this;
+         }
+
+         /**
+          * Returns the value that the iterator points to.
+          * @ingroup keyvalue
+          *
+          * @return The value that the iterator points to.
+          */
+         T value() const {
+            using namespace detail;
+
+            eosio::check(itr_stat != kv_it_stat::iterator_end, "Cannot read end iterator");
+
+            uint32_t value_size;
+            uint32_t actual_value_size;
+            uint32_t offset = 0;
+
+            // call once to get the value_size
+            internal_use_do_not_use::kv_it_value(itr, 0, (char*)nullptr, 0, value_size);
+
+            void* buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+            auto stat = internal_use_do_not_use::kv_it_value(itr, offset, (char*)buffer, value_size, actual_value_size);
+
+            eosio::check(static_cast<kv_it_stat>(stat) != kv_it_stat::iterator_end, "Error reading value");
+
+            void* serialize = buffer;
+            size_t serialize_size = actual_value_size;
+
+            if (idx->name != idx->tbl->primary_index->name) {
+               uint32_t actual_data_size;
+               auto success = internal_use_do_not_use::kv_get(idx->tbl->db_name, contract_name.value, (char*)buffer, actual_value_size, actual_data_size);
+               eosio::check(success, "failure getting primary key");
+
+               void* pk_buffer = actual_data_size > detail::max_stack_buffer_size ? malloc(actual_data_size) : alloca(actual_data_size);
+               auto copy_size = internal_use_do_not_use::kv_get_data(idx->tbl->db_name, 0, (char*)pk_buffer, actual_data_size);
+
+               eosio::check(copy_size != actual_value_size, "failure getting primary index data");
```

Comment:
I'm not sure what you're trying to test here.  Won't this fail when the primary key has the same size as its value?

Reply:
Yeah I think this is an artifact of old code (that looks like it might have been wrong originally anyways). I don't think the intent of this test is really needed.